src/internal/packager/images/pull.go

```diff
@@ -27,329 +32,329 @@ import (
 
-			metadataImageConcurrency.ProgressChan <- ImgInfo{RefInfo: refInfo, Img: img, HasImageLayers: hasImageLayers}
-		}()
-	}
-
-	onMetadataProgress := func(finishedImage ImgInfo, iteration int) {
-		spinner.Updatef("Fetching image metadata (%d of %d): %s", iteration+1, len(i.ImageList), finishedImage.RefInfo.Reference)
-		refInfoToImage[finishedImage.RefInfo] = finishedImage.Img
-		imgInfoList = append(imgInfoList, finishedImage)
-	}
+			if refInfo.Digest != "" && desc != nil && types.MediaType(desc.MediaType).IsIndex() {
+				message.Warn("Zarf does not currently support direct consumption of OCI image indexes or Docker manifest lists")
 
-	onMetadataError := func(err error) error {
-		return err
-	}
+				var idx v1.IndexManifest
+				if err := json.Unmarshal(desc.Manifest, &idx); err != nil {
+					return fmt.Errorf("unable to unmarshal index manifest: %w", err)
+				}
+				lines := []string{"The following images are available in the index:"}
+				name := refInfo.Name
+				if refInfo.Tag != "" {
+					name += ":" + refInfo.Tag
+				}
+				for _, desc := range idx.Manifests {
+					lines = append(lines, fmt.Sprintf("\n(%s) %s@%s", desc.Platform, name, desc.Digest))
+				}
+				message.Warn(strings.Join(lines, "\n"))
+				cancel()
+				return fmt.Errorf("%s resolved to an index, please select a specific platform to use", refInfo.Reference)
+			}
 
-	if err := metadataImageConcurrency.WaitWithProgress(onMetadataProgress, onMetadataError); err != nil {
-		return nil, err
-	}
+			img = cache.Image(img, cache.NewFilesystemCache(cfg.CacheDirectory))
 
-	// Create the ImagePath directory
-	if err := helpers.CreateDirectory(i.ImagesPath, helpers.ReadExecuteAllWriteUser); err != nil {
-		return nil, fmt.Errorf("failed to create image path %s: %w", i.ImagesPath, err)
-	}
+			manifest, err := img.Manifest()
+			if err != nil {
+				return fmt.Errorf("unable to get manifest for %s: %w", refInfo.Reference, err)
+			}
+			totalBytes.Add(manifest.Config.Size)
 
-	totalBytes := int64(0)
-	processedLayers := make(map[string]v1.Layer)
-	for refInfo, img := range refInfoToImage {
-		// Get the byte size for this image
-		layers, err := img.Layers()
-		if err != nil {
-			return nil, fmt.Errorf("unable to get layers for image %s: %w", refInfo.Reference, err)
-		}
-		for _, layer := range layers {
-			layerDigest, err := layer.Digest()
+			layers, err := img.Layers()
 			if err != nil {
-				return nil, fmt.Errorf("unable to get digest for image layer: %w", err)
+				return fmt.Errorf("unable to get layers for %s: %w", refInfo.Reference, err)
 			}
 
-			// Only calculate this layer size if we haven't already looked at it
-			if _, ok := processedLayers[layerDigest.Hex]; !ok {
-				size, err := layer.Size()
+			for _, layer := range layers {
+				digest, err := layer.Digest()
 				if err != nil {
-					return nil, fmt.Errorf("unable to get size of layer: %w", err)
+					return fmt.Errorf("unable to get digest for image layer: %w", err)
+				}
+
+				shaLock.Lock()
```

Comment:
Lock is not released on defer here. If an error is returned due to the layer size the lock will never be released. Any other pull waiting for the critical section will block forever.